pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package deferunlockcheck
+
+import (
+	"go/ast"
+
+	"github.com/cockroachdb/cockroach/pkg/testutils/lint/passes/passesutil"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+// Doc documents this pass.
+const Doc = "checks that usages of mutex Unlock() are deferred."
+
+const noLintName = "deferunlock"
+
+// Analyzer is an analysis pass that checks for mutex unlocks which
+// aren't deferred.
+var Analyzer = &analysis.Analyzer{
+	Name:     "deferunlockcheck",
+	Doc:      Doc,
+	Requires: []*analysis.Analyzer{inspect.Analyzer},
+	Run:      run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	astInspector := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	// Note: defer ...Unlock() expressions are captured as ast.DeferStmt nodes so we don't have to worry about those results returning
+	// for ast.ExprStmt nodes.
+	filter := []ast.Node{
+		(*ast.ExprStmt)(nil),
+	}
+	astInspector.Preorder(filter, func(n ast.Node) {
+		stmt := n.(*ast.ExprStmt)
+		expr, ok := stmt.X.(*ast.CallExpr)
+		if !ok {
+			return
+		}
+		sel, ok := expr.Fun.(*ast.SelectorExpr)
+		if !ok {
+			return
+		}
+		if sel.Sel != nil && (sel.Sel.Name == "Unlock" || sel.Sel.Name == "RUnlock") && !passesutil.HasNolintComment(pass, n, noLintName) {
```

Comment:
We probably want to make sure we're only reporting issues with `Unlock` calls called on instances of `sync{,util}.{RW}Mutex`. Currently if a custom struct implements an `Unlock` function, the linter will complain about it and we probably don't want that, as the semantics are unknown to the linter.

Some similar code:

https://github.com/cockroachdb/cockroach/blob/75c6056aad3574afef45888533c3df9133ddf65d/pkg/testutils/lint/passes/loopvarcapture/loopvarcapture.go#L404-L407